### DISCUSSION TITLE
I made a small script to fix Sunshine's monitor assignment automatically

### DISCUSSION THREAD

**Discussion by alydevs** [LizardByte/community#434] [Sunshine]
I use Sunshine and Moonlight every day to repurpose my laptop as a second monitor for my desktop computer. I have an HDMI cable going to the room TV, when I use the TV I turn it on, but otherwise the laptop is showing the TV's monitor. This has so many advantages other than just providing an extra monitor on my desk, mainly that the laptop also passes through all Bluetooth devices including headphones and kb/m with minimal latency, as the desktop motherboard doesn't have wireless support.

The issue is that if I change the screen layout, set it to Duplicate or Display only on x, the \\DISPLAYn number changes and therefore Sunshine loses track of it. The other day I installed a GPU driver update and the numbers changed to DISPLAY6 and DISPLAY7 for some reason; usually though the change is limited to DISPLAY1 and DISPLAY2 switching around. You'd expect the main monitor to be DISPLAY1, but currently it's DISPLAY2.

So I made this script that I can run when Sunshine is pointing to the wrong or invalid screen, which fixes it to point to the 1080p screen, as opposed to the 1200p main monitor.

edit: I've updated this to now check if the display name changes, and also allow selecting display by serial number. Save the PowerShell script at the bottom to monitor.ps1 in the same directory as the Python script.

```python
from requests import post, get
from requests.exceptions import ConnectionError
from subprocess import check_output
from json import loads

SN = "" # run monitor.ps1 once to find your desired monitor's serial number
PASSWORD = "" # your Sunshine web ui password

dxgi = check_output(["powershell.exe", r'.\monitor.ps1']).decode("utf-8")
try:
    display = [r"\\.\DISPLAY" + str(d["DisplayIdentity"]) for d in loads(dxgi) if d["SerialNumber"] == SN][0]
    current = get("https://localhost:47990/api/config", auth=("sunshine", PASSWORD), verify=False).json().get("output_name")
    if display != current:
        print("Updating display variable from {} to {}".format(current, display))
        post("https://localhost:47990/api/config", json={"output_name":display}, auth=("sunshine", PASSWORD), verify=False)
        post("https://localhost:47990/api/restart", auth=("sunshine", PASSWORD), verify=False)
except IndexError:
    print("Display not found")
    print(dxgi)
except ConnectionError:
    pass
```

```powershell
# credit to https://old.reddit.com/r/PowerShell/comments/jmtxpx/multiple_monitors_identifying_and_manipulating/gay0v3o/
Try
{
    $monitorsRegistry = Get-Item "HKLM:\SYSTEM\CurrentControlSet\Services\monitor\Enum"
    $monitorsRegistryValueNames = $monitorsRegistry.GetValueNames() | where { $_ -match '\d' }
    $monitors_RegistryInfo = foreach ($monitorsRegistryValueName in $monitorsRegistryValueNames)
    {
        [pscustomobject]@{
            Display  = $monitorsRegistryValueName
            DeviceID = $monitorsRegistry.GetValue("$monitorsRegistryValueName")
        }
    }

    #Servers - NOT SUPPORTED
    $monitors_wmimonitorid = Get-CIMInstance -Query "Select * FROM WMIMonitorID" -Namespace root\wmi -ErrorAction SilentlyContinue | Select `
        @{ n = "InstanceName"; e = { $_.InstanceName.Split("_")[0] } },
        @{ n = "SerialNumber"; e = { [System.Text.Encoding]::ASCII.GetString($_.SerialNumberID -ne 00) }     },`
        YearOfManufacture
    Add-Type -AssemblyName System.Windows.Forms
    $Monitors = [System.Windows.Forms.Screen]::AllScreens
    $monitors_AllScreens = foreach ($Monitor in $Monitors)
    {
        [pscustomobject]@{
            DeviceName = ($Monitor.DeviceName.Split('Y')[-1]) - 1
            Primary    = $Monitor.Primary
            BitsPerPixel = $Monitor.BitsPerPixel
            Width      = $Monitor.bounds.Width
            Height     = $Monitor.bounds.Height
            WxH        = "$($Monitor.bounds.Width) x $($Monitor.bounds.Height)"
            X          = $Monitor.bounds.X
            Y          = $Monitor.bounds.Y
            Location   = "$($Monitor.bounds.X) x $($Monitor.bounds.Y)"
        }
    }
    $allMonitorInfo = foreach ($monitor_RegistryInfo in $monitors_RegistryInfo)
    {
		$monitor_AllScreens = $monitors_AllScreens | where { $_.DeviceName -eq $monitor_RegistryInfo.Display }
        $monitor_wmimonitorid = $monitors_wmimonitorid | where { $_.InstanceName -eq $monitor_RegistryInfo.DeviceID }

        [pscustomobject]@{
            DisplayIdentity = ([int]$monitor_RegistryInfo.Display) + 1
            SerialNumber = $monitor_wmimonitorid.SerialNumber
            BoundsWidth  = $monitor_AllScreens.Width
            BoundsHeight = $monitor_AllScreens.Height
            BoundsX      = $monitor_AllScreens.X
            BoundsY      = $monitor_AllScreens.Y
        }
    }
    $allMonitorInfo | ConvertTo-Json | Write-Output
}
Catch
{
    $_.Exception.Message
}
```

My next plan is probably to ~~set this up to run automatically~~, ~~check if the display Sunshine is pointing to is an invalid screen~~, and ~~fix it when there are no clients connected, pointing to the 1200p main screen as a last resort~~.

I hope this is helpful for others who might want to use Sunshine and Moonlight to repurpose otherwise-unused hardware for their setup.

**Comment by Laymer** [comment]
@alyssadev This is very helpful ! I use sunlight to stream to the living room TV **BUT** since I often need to rearrange my displays (+had to reset CMOS recently), my config was invalid. Great idea to have it in a script ! Thanks for sharing 🤓

**Reply by alydevs** [reply]
I've updated the script to (with the help of a Powershell script to retrieve the information easier than with Python) allow selecting display by serial number.

[a comment by an automated account is not shown]
